### ISSUE TITLE
Add Support for Lazy Request Injection

### ISSUE THREAD

**Issue by StarLard** [kean/FetchImage#7]
### Problem

Currently, `FetchImage` requires the URL to be known at initialization time. This is an issue because it means no view can declare ownership over the `FetchImage` object via `StateObject`. SwiftUI views' initializers may be called multiple times meaning that your `FetchImage` will be constructed multiple times (thus executing fetch multiple times) unless it can be declared outside the initializer via a `StateObject` property wrapper.

The current API also means requires you to declare a new instance of `FetchImage` if your URL ever changes. Ideally, we'd be able to just modify the `FetchImage.request` property and call `fetch` again.

Desired usage:
```
struct ContentView: View {
    @StateObject
    let image = FetchImage()

    var body: some View {
        ...
    }

    init(url: URL) {
        self.image.request = ImageRequest(url: url)
    }
}
```

### Proposed Solution

Allow URL to be `nil` at init time and inject it lazily:

```
public final class FetchImage: ObservableObject, Identifiable {
    /// The original request.
    ///
    /// Modifying this property while a regular image quality request is in-flight will cancel the request. Setting
    /// it to `nil` while the regular quality image is loaded will clear the current image.
    ///
    /// - note: Only modify this property from the main thread.
    public var request: ImageRequest? {
        didSet {
            assert(Thread.isMainThread, "Only modify the request from the main thread.")
            if currentlyLoadingImageQuality == .regular {
                cancel()
            }
            guard let newRequest = request else {
                if loadedImageQuality == .regular {
                    image = nil
                }
                return
            }
            priority = newRequest.priority
        }
    }

    /// The request to be performed if the original request fails with
    /// `networkUnavailableReason` `.constrained` (low data mode).
    ///
    /// Modifying this property while a low image quality request is in-flight will cancel the request. Setting
    /// it to `nil` while the low quality image is loaded will clear the current image.
    ///
    /// - note: Only modify this property from the main thread.
    public var lowDataRequest: ImageRequest? {
        didSet {
            assert(Thread.isMainThread, "Only modify the request from the main thread.")
            if currentlyLoadingImageQuality == .low {
                cancel()
            }
            if lowDataRequest == nil && loadedImageQuality == .low {
                image = nil
            }
        }
    }

    /// Returns the fetched image.
    ///
    /// - note: In case pipeline has `isProgressiveDecodingEnabled` option enabled
    /// and the image being downloaded supports progressive decoding, the `image`
    /// might be updated multiple times during the download.
    @Published public private(set) var image: PlatformImage?

    /// Returns an error if the previous attempt to fetch the image failed with an error.
    /// Error is cleared out when the download is restarted.
    @Published public private(set) var error: Error?

    /// Returns `true` if the image is being loaded.
    @Published public private(set) var isLoading: Bool = false

    public struct Progress {
        /// The number of bytes that the task has received.
        public let completed: Int64

        /// A best-guess upper bound on the number of bytes the client expects to send.
        public let total: Int64
    }

    /// The progress of the image download.
    @Published public var progress = Progress(completed: 0, total: 0)

    /// Updates the priority of the task, even if the task is already running.
    public var priority: ImageRequest.Priority {
        didSet { task?.priority = priority }
    }

    private let pipeline: ImagePipeline
    private var task: ImageTask?
    private var loadedImageQuality: ImageQuality? = nil
    private var currentlyLoadingImageQuality: ImageQuality? = nil

    private enum ImageQuality {
        case regular, low
    }

    deinit {
        cancel()
    }

    /// Initializes the fetch request and immediately start loading.
    public init(request: ImageRequest? = nil, lowDataRequest: ImageRequest? = nil, pipeline: ImagePipeline = .shared) {
        self.request = request
        self.lowDataRequest = lowDataRequest
        self.priority = request?.priority ?? .normal
        self.pipeline = pipeline

        self.fetch()
    }

    /// Initializes the fetch request and immediately start loading.
    public convenience init(url: URL, pipeline: ImagePipeline = .shared) {
        self.init(request: ImageRequest(url: url), pipeline: pipeline)
    }

    /// A convenience initializer that fetches the image with a regular URL with
    /// constrained network access disabled, and if the download fails because of
    /// the constrained network access, uses a low data URL instead.
    public convenience init(regularUrl: URL, lowDataUrl: URL, pipeline: ImagePipeline = .shared) {
        var request = URLRequest(url: regularUrl)
        request.allowsConstrainedNetworkAccess = false

        self.init(request: ImageRequest(urlRequest: request), lowDataRequest: ImageRequest(url: lowDataUrl), pipeline: pipeline)
    }

    /// Starts loading the image if not already loaded and the download is not
    /// already in progress.
    ///
    /// - note: Low Data Mode. If the `lowDataRequest` is provided and the regular
    /// request fails because of the constrained network access, the fetcher tries
    /// to download the low-quality image. The fetcher always tries to get the high
    /// quality image. If the first attempt fails, the next time you call `fetch`,
    /// it is going to attempt to fetch the regular quality image again.
    public func fetch() {
        guard !isLoading, loadedImageQuality != .regular, let request = self.request else {
            return
        }

        error = nil

        // Try to display the regular image if it is available in memory cache
        if let container = pipeline.cachedImage(for: request) {
            (image, loadedImageQuality) = (container.image, .regular)
            return // Nothing to do
        }

        // Try to display the low data image and retry loading the regular image
        if let container = lowDataRequest.flatMap(pipeline.cachedImage(for:)) {
            (image, loadedImageQuality) = (container.image, .low)
        }

        isLoading = true
        loadImage(request: request, quality: .regular)
    }

    private func loadImage(request: ImageRequest, quality: ImageQuality) {
        progress = Progress(completed: 0, total: 0)
        currentlyLoadingImageQuality = quality
        task = pipeline.loadImage(
            with: request,
            progress: { [weak self] response, completed, total in
                guard let self = self else { return }

                self.progress = Progress(completed: completed, total: total)

                if let image = response?.image {
                    self.image = image // Display progressively decoded image
                }
            },
            completion: { [weak self] in
                self?.didFinishRequest(result: $0, quality: quality)
            }
        )

        if priority != request.priority {
            task?.priority = priority
        }
    }

    private func didFinishRequest(result: Result<ImageResponse, ImagePipeline.Error>, quality: ImageQuality) {
        task = nil
        currentlyLoadingImageQuality = nil
        switch result {
        case let .success(response):
            isLoading = false
            (image, loadedImageQuality) = (response.image, quality)
        case let .failure(error):
            // If the regular request fails because of the low data mode,
            // use an alternative source.
            if quality == .regular, error.isConstrainedNetwork, let request = self.lowDataRequest {
                if loadedImageQuality == .low {
                    isLoading = false // Low-quality image already loaded
                } else {
                    loadImage(request: request, quality: .low)
                }
            } else {
                self.error = error
                isLoading = false
            }
        }
    }

    /// Marks the request as being cancelled.
    public func cancel() {
        task?.cancel() // Guarantees that no more callbacks are will be delivered
        task = nil
        isLoading = false
    }
}
```

**Comment by StarLard** [comment]
An added bonus of this approach is it allows you to defer calling `fetch` until `onAppear`. This is particularly useful for lazy stacks in which there may be many elements.

**Comment by kean** [comment]
Hey, @StarLard. Thanks for putting this together. I like the direction, I just need to review it a bit more thoroughly when I get some free time.

**Comment by kean** [comment]
There is another problem I would like to address. Currently `FetchImage` retains an image. This is a problem for long lists. Only image views that are currently on screen and in-memory caches should retain images.

**Comment by kean** [comment]
Hey, I'm going to add deferred loading in the upcoming release.

As for `@StateObject`, it seems what you want is initializer injection. You can do that the following way:

```
struct ImageView: View {
    @StateObject var image: FetchImage

    init(url: URL) {
        _image = StateObject(wrappedValue: FetchImage(url: url))
    }
}
```

Not ideal, but better adding unnecessary mutation to `FetchImage` just because `@StateObject` sucks.

**Comment by kean** [comment]
There is even a clear way to do that:

```swift
struct ContentView: View {
    var body: some View {
        List(demoPhotosURLs, id: \.self) { url in
            ImageView(image: FetchImage(url: url))
        }
    }
}

public struct ImageView: View {
    @StateObject var image: FetchImage
}
```

**Comment by StarLard** [comment]
Thanks for the response @kean; however, I think this issue should be re-opened. The example solution you provided creates a bug. SwiftUI view initializers may actually be called multiple times for the same view, meaning that in your example your `StateObject` is also potentially being initialized multiple times; thus negating the benefits of using `StateObject`. Xcode will actually warn you about this at runtime when it notices that the `StateObject` has been initialized multiple times.

**Comment by kean** [comment]
I knew I shouldn't have trusted a random advice online without fully verifying it.

Turns out StateObject is a pretty bad API because of the lack of initializer injection. In [data essentials](https://developer.apple.com/videos/play/wwdc2020/10040/) they recommend passing a URL in the `load` method. I'm not against this approach, but I'm a bit disappointed because then `FetchImage` becomes essentially becomes [`ImageViewController`](https://github.com/kean/Nuke/blob/master/Sources/ImageViewExtensions.swift#L313) which I built for UIKit. We are not getting rid of any of the state / mutation.

**Comment by StarLard** [comment]
I like your suggestion of passing the URL in `load`. Yeah `FetchImage` ultimately becomes a kind of view model for a generic image view. If we want to preserve the current stateless API of `FetchImage`, I think we would need to provide a concrete view implementation that uses `FetchImage` and handles the state management there.

**Comment by StarLard** [comment]
Happy to put up a PR for this if you would like 👍

**Comment by kean** [comment]
Yeah. I'm not actively working with SwiftUI, I would appreciate help. I'm thinking would it be possible to support all the use-cases in a clear way: @ObservedObject and @StateObject. Your initial proposal seem to do both. The approach suggested by Apple should also work.

**Comment by kean** [comment]
I tested the `image = StateObject(wrappedValue: FetchImage(url: url))` suggestion and it seems to work perfectly fine https://gist.github.com/kean/838cad5c574a2765ea45decf6e4b1b23

**Comment by StarLard** [comment]
I think the issue lies in re-initializing the `StateObject`, not the wrapped value. As you can see from your example, the subview init is being called multiple times, thus the StateObject initializer is also being called multiple times. I'm not sure what the side effects of this all may be, but it does state in the [documentation](https://developer.apple.com/documentation/swiftui/stateobject/init(wrappedvalue:)) that the `StateObject` initializer should never be called directly. Your example isn't reproducing the runtime warning, but you will trigger it if you attempt to access the wrapped value in the initialzer (you should see the "Accessing StateObject's object without being installed on a View. This will create a new instance each time." error in the console). According to Apple, StateObject is intended to initialized by the system just before your view's body is called.

**Comment by kean** [comment]
> initializer should never be called directly
It doesn't say *never*, it just indicates what the expected usage is.

> StateObject initializer is also being called multiple times
That also happens if you declare it the "correct" way. It's just a struct. Whenever `Subview` in my example is created, `StateObject` also is. But the `wrappedValue`, which is a closure, is only called once per "instance" of a view. I'm fairly confident the is how it works because I did some [research](https://kean.blog/post/swiftui-data-flow) into this. There is no compiler magic in SwiftUI when it comes to property wrappers, it's all done in runtime using reflection.

> if you attempt to access the wrapped value in the initialzer 
Yes, this you should never do. This also includes other property wrappers such as State. You should only every access the values in the body. SwiftUI requires an additional pass after initialization to associate property wrappers with views.

I would recommend to just call the initializer directly as many people suggest. It's not a hack, it uses public APIs and documented behavior.

**Comment by kean** [comment]
I confirmed with some folks from Apple that using `StateObject` initializer this way is perfectly fine. It's not ergonomic aka the code doesn't look great. But it works as expected.

**Comment by StarLard** [comment]
Thanks for digging in on this and verifying the correct usage with Apple! I think this issue is good to be closed now.

**Comment by eliottrobson** [comment]
Changing `@ObservedObject` to `@StateObject` for the FetchImage resolved a lot of issues for me when the views were created dynamically (from fetch request). I was getting flickering views that sometimes disappeared, this solved it immediately. I was mid-way through trying to build a simple repro and found this issue. Is it something that can be called out in the readme to help others?

**Comment by davut** [comment]
`@StateObject` seems to work perfectly but I need to support iOS 13+ is there any way to use it with `@ObservedObject`?

**Comment by kean** [comment]
@davut with `@ObservedObject`, you need to make sure you create your `FetchImage` objects once and inject them. I'm not sure what's the best way to achieve that without `@StateObject`. I only have one basic SwiftUI app and it has a hierarchy of view models.

**Comment by kean** [comment]
Reopening. `@StateObject` approach recommended on WWDC appears to be the least worst option. PR with suggested changes https://github.com/kean/FetchImage/pull/14/files. I haven't tested it yet and I'm not using FetchImage in production, so I'm not 100% sure about it. @StarLard I would appreciate your input on this.

**Comment by kean** [comment]
First class `@StateObject` support now available in [FetchImage 0.4.0](https://github.com/kean/FetchImage/releases/tag/0.4.0)
